mono/metadata/object.c

```diff
@@ -3955,36 +3955,46 @@ mono_runtime_get_main_args (void)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 	MonoError error;
-	MonoArray *result = mono_runtime_get_main_args_checked (&error);
+	MonoArrayHandle arg_array = mono_runtime_get_main_args_handle (&error);
+	MonoArray *result = MONO_HANDLE_RAW (arg_array);
 	mono_error_assert_ok (&error);
 	return result;
 }
 
+static MonoStringHandle
+string_from_main_arg (MonoDomain *domain, int idx, MonoError *error)
+{
+	HANDLE_FUNCTION_ENTER ();
+	error_init (error);
+	MonoStringHandle res = mono_string_new_handle (domain, main_args [idx], error);
+	HANDLE_FUNCTION_RETURN_VAL (res);
```

Comment:
See other comments... `HANDLE_FUNCTION_RETURN_VAL()` isn't the right thing to use with functions that return handles.  Also it's better if this helper assigns into the destination array, instead of returning a value.